### PR TITLE
Do not treat certain calls as intrinsics for RyuJIT/x86.

### PR DESCRIPTION
**Pull request by pgavlin**
On x86 RyuJIT, importing intrinsics that are implemented as user calls
can cause incorrect calculation of the depth of the stack if these
intrinsics are used as arguments to another call. This causes bad code
generation for certain EH constructs. Instead of implementing these
intrinsics as intrinsics, simply import them as calls.

### CODE REVIEW

**Comment by pgavlin** [comment]
@BruceForstall @dotnet/jit-contrib PTAL.

Fixes VSO 280127.

**Comment by pgavlin** [comment]
Ping.

**Comment by adiaaida** [comment]
LGTM

File: src/jit/importer.cpp [hunk 1]
```diff
@@ -3344,16 +3344,21 @@ GenTreePtr Compiler::impIntrinsic(CORINFO_CLASS_HANDLE  clsHnd,
 
             op1 = nullptr;
 
-#ifdef LEGACY_BACKEND
+#if defined(LEGACY_BACKEND)
             if (IsTargetIntrinsic(intrinsicID))
-#else
+#elif !defined(_TARGET_X86_)
```
Comment (pgavlin) [inline src/jit/importer.cpp]: Yes: it is done during morph.

File: src/jit/importer.cpp [hunk 2]
```diff
@@ -3344,16 +3344,21 @@ GenTreePtr Compiler::impIntrinsic(CORINFO_CLASS_HANDLE  clsHnd,
...
             // don't do this optimization, because
             //  a) For back compatibility reasons on desktop.Net 4.6 / 4.6.1
             //  b) It will be non-trivial task or too late to re-materialize a surviving
             //     tail prefixed GT_INTRINSIC as tail call in rationalizer.
             if (!IsIntrinsicImplementedByUserCall(intrinsicID) || !tailCall)
+#else
+            // On x86 RyuJIT, importing intrinsics that are implemented as user calls can cause incorrect calculation
+            // of the depth of the stack if these intrinsics are used as arguments to another call. This causes bad
+            // code generation for certain EH constructs.
+            if (!IsIntrinsicImplementedByUserCall(intrinsicID))
```
Comment (pgavlin) [inline src/jit/importer.cpp]: Already done: #8243.

I think the most straightforward (if brittle) fix is probably to put together a table from intrinsic -> stack depth changes and consult this during morph in order to properly calculate the stack depth.

Comment (CarolEidt) [inline src/jit/importer.cpp]: I think that we should be able to deal with this in the rationalizer at the time it changes them back into user calls. As you say, it is probably not the most straightforward fix, but is probably a better way to handle it. It's a bit unfortunate to be dealing with stack depth so early in the JIT.

---
In reply to: [89024918](https://github.com/dotnet/coreclr/pull/8235#discussion_r89024918) [](ancestors = 89024918)

File: src/jit/importer.cpp (hunk 1 above)
Comment (CarolEidt) [inline src/jit/importer.cpp]: I second Siva's suggestion.

**Pull request merged**